platform/eventing/listener.go

```diff
@@ -0,0 +1,74 @@
+package eventing
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/cloudevents/sdk-go/v2/client"
+	"github.com/cloudevents/sdk-go/v2/event"
+	cloudeventsnats "github.com/cloudevents/sdk-go/v2/protocol/nats"
+	"github.com/jmoiron/sqlx"
+)
+
+type listenerSubject string
+
+// Listener defines what a nats listener looks like
+type Listener struct {
+	URL      string
+	subjects []string
+	db       *sqlx.DB
+	logger   *log.Logger
+}
+
+// NewListener creates a new config for a listener
+func NewListener(URL string, subjects []string, db *sqlx.DB, logger *log.Logger) *Listener {
+	return &Listener{
+		URL:      URL,
+		subjects: subjects,
+		db:       db,
+		logger:   logger,
+	}
+}
+
+// Start starts a listener for the provided subjects
+// It will save all events for the subjects in the DB
+func (listener *Listener) Start() {
```

Comment:
I'm not really comfortable with this function as it spawn several goroutines but, if some of them can't connect or crash in the middle of their loop, we won't be notified. It means we will just miss a ton of messages.

I think we should either monitor the goroutines and retries the ones that crash or crash them all and restart the whole application. 

But anyway I don't think we should silently let them crash. When I say silently I mean you log but logs are not intended to trigger alerts or be read in a day to day basis. 

So we might stay days without receiving some events and have weird behaviour because of this.

Reply:
Because the `natsClient.StartReceiver` call is blocking, it won't always be checking if something happened to the listeners, only when it returns an error. 

Therefore I went with setting a max number of retries via the config and logging to Sentry in case it fails and just restart that listener. 

I think with this approach we can still closely monitor if things go wrong somewhere without letting it go a-wohl. What do you think about this?

Reply:
I'm not quite sure. What do you think of a small pair programming session to tackle this together?